### PR TITLE
Implement PR #1682 - TPA Event

### PR DESCRIPTION
**Pull request by DrewAMitchell**
https://github.com/EssentialsX/Essentials/issues/1682
First pass at this solution, awaiting feedback. I've implemented the event, but was unsure what to do in the event that it was cancelled, so I made it to where neither party of the TPA request will receive any messages.
Another potential solution is only have the requester receive a message, similar to if they're being ignored.
Also was unsure of what to do for TPAAll, so give me feedback on that as well.

### CODE REVIEW

File: Essentials/src/com/earth2me/essentials/commands/Commandtpa.java
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpa.java]: format

Comment (drtshock) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpa.java]: Also, why are we constructing the event without knowing if we're going to use it for sure or not?

File: Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java]: format again

File: Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java:47]: line spacing..

**Review by drtshock — CHANGES_REQUESTED** [review]
On top of what's in my comments, what's the reason for not extending one of our current events? https://github.com/EssentialsX/Essentials/blob/2.x/Essentials/src/net/ess3/api/events/FlyStatusChangeEvent.java

File: Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java]: I think that a plugin should let the user know if it cancelled the plugin and why, though EssentialsX *could* output a generic "Your teleport request was cancelled." message - I'd need to see how other EssentialsX events do this.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (mdcfe) [inline Essentials/src/net/ess3/api/events/TPARequestEvent.java]: EssentialsX does not use tab indents.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (mdcfe) [inline Essentials/src/net/ess3/api/events/TPARequestEvent.java]: This method name is long and confusing. The `RequestTo` part is redundant since the whole event is regarding the event.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (mdcfe) [inline Essentials/src/net/ess3/api/events/TPARequestEvent.java]: We don't need both of these methods either - I'd suggest renaming `isRequestToTeleportHere` to `isTeleportHere` and removing `isRequestToTeleportTo`.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (mdcfe) [inline Essentials/src/net/ess3/api/events/TPARequestEvent.java]: I'd suggest renaming this to `target` as it seems more in line with other references to the person receiving the request.

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java
Comment (DrewAMitchell) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java]: Will wait for further advice on this, willing to go either way.

File: Essentials/src/com/earth2me/essentials/commands/Commandtpa.java
Comment (DrewAMitchell) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpa.java]: Apologies, aforementioned issues will be fixed in the next commit.

File: Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java
Comment (DrewAMitchell) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java:47]: It was caused by tab indentation, will fix in the next commit.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (DrewAMitchell) [reply Essentials/src/net/ess3/api/events/TPARequestEvent.java]: Dually noted, will be converted in next commit.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (DrewAMitchell) [reply Essentials/src/net/ess3/api/events/TPARequestEvent.java]: Done!

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (DrewAMitchell) [reply Essentials/src/net/ess3/api/events/TPARequestEvent.java]: Absolutely. Will be fixed in the next commit.

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (DrewAMitchell) [reply Essentials/src/net/ess3/api/events/TPARequestEvent.java]: Taken into account!

File: Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java
Comment (DrewAMitchell) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java]: I hope this isn't a line spacing problem - I'll fix it in a future commit if it is, just review this next commit and let me know if it's still there.

**Comment by DrewAMitchell** [comment]
@drtshock I didn't use one of the current base events because I figured that the getRequester could be checked if it was an instanceof IUser, as the console command sender can trigger this event as well, so I made the requester a CommandSource.

**Comment by khobbits** [comment]
I think I was just being lazy when I wrote StateChangeEvent, the reason was that there could be a dozen different triggers for commands, once you start including mods, console, rpc.  Easier to just work on the idea that if it was a player, we'd have a valid IUser, if not, null was as useful as anything else.

File: Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java
Comment (DrewAMitchell) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java]: Commit will include a teleport request cancelled message, unless you've got another recommendation.

**Comment by DrewAMitchell** [comment]
@md678685 @drtshock @SupaHam Re-review please!

**Comment by DrewAMitchell** [comment]
My recent commit is still awaiting review; I will absolutely remove any TODO comments as requested. They're just there to allow me to remember what I wanted reviewed most from you guys. XD

File: Essentials/src/messages_fr.properties
Comment (mdcfe) [inline Essentials/src/messages_fr.properties]: Could you readd the new line at the end of this file?

File: Essentials/src/messages_it.properties
Comment (mdcfe) [inline Essentials/src/messages_it.properties]: Same with newline here.

File: Essentials/src/messages_pl.properties
Comment (mdcfe) [inline Essentials/src/messages_pl.properties]: Again, newline here.

File: Essentials/src/messages_zh.properties
Comment (mdcfe) [inline Essentials/src/messages_zh.properties]: Last missing newline in a language file.

File: Essentials/src/messages_cs.properties
Comment (mdcfe) [inline Essentials/src/messages_cs.properties]: I missed another missing newline here.

File: Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java]: Any particular reason to reorder the imports, or was this just an IDE thing?

File: Essentials/src/net/ess3/api/events/TPARequestEvent.java
Comment (mdcfe) [inline Essentials/src/net/ess3/api/events/TPARequestEvent.java:10]: @drtshock @SupaHam Thoughts on having a `showCancellationMessage` boolean defaulting to `true` that determines whether EssentialsX shows its own "request cancelled" message, in case a plugin wants to display its own message instead?

**Review by mdcfe — CHANGES_REQUESTED** [review]
There's a couple of really minor things like certain files being accidentally altered (presumably by the IDE), and I'm awaiting feedback from the others about whether to allow plugins to disable the cancellation message, but otherwise, this looks good.

File: Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java
Comment (DrewAMitchell) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java]: IDE thing, I'll fix it if you'd like me to - might fix it just because it's unnecessary.  :)

File: Essentials/src/messages_cs.properties
Comment (DrewAMitchell) [reply Essentials/src/messages_cs.properties]: So, to clarify, you want an empty line at the end of the lang files?

**Comment by DrewAMitchell** [comment]
I will be awaiting final thoughts before I make changes to the commit in the way of the new lines.
For clarification,
```yaml
someLanguageString="Test String; this is an example, so it doesn't matter."
#empty line
```
That empty line should be at the bottom of all the language files? Or is it just a cr/lf issue?

**Comment by mdcfe** [comment]
I'm pretty sure the newlines are supposed to be at the bottom of the file, but you can leave them as-is for now.

Feel free to revert the changes in the order of imports back to how they were before - it would minimise the diff.

Still awaiting feedback from the others on my suggestion though.

**Comment by DrewAMitchell** [comment]
@md678685 You'll be happy to know that the latest squashed commit keeps your precious new-lines and minimizes the diff with the imports. I am now solely waiting on feedback from your companions.
@drtshock @SupaHam

**Comment by mdcfe** [comment]
@drtshock @SupaHam Should we allow other plugins to manually disable our "request cancelled" message in case they want to show their own message?

**Comment by SupaHam** [comment]
I don't think so. It'll be fine to have it print the official message and
the plugin to print theirs if necessary.

**Review by mdcfe — DISMISSED** [review]

**Comment by DrewAMitchell** [comment]
@md678685 I'll remove my TODO's and comments and all that.

**Review by drtshock — DISMISSED** [review]

**Review by drtshock — DISMISSED** [review]

**Review by mdcfe — DISMISSED** [review]

File: Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java]: Is this TODO still applicable? I think the initial message should be sent regardless of whether some are cancelled.

**Review by mdcfe — COMMENTED** [review]
This is pretty much ready to merge, pending a decision on the TODO.

**Review by Ichbinjoe — DISMISSED** [review]

**Comment by Ichbinjoe** [comment]
Looking for someone else to sign off on this (because I made changes)  then I'll merge this in.

File: Essentials/src/messages_cs.properties
Comment (mdcfe) [inline Essentials/src/messages_cs.properties]: Since the Crowdin integration, we now need to leave the translation files alone and only put new messages in `messages.properties`, else we'll get merge conflicts.

Edit: I'll fix this when I get time, but if anyone wants to fix this in the meantime feel free.

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: Essentials/src/com/earth2me/essentials/commands/Commandtpa.java
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpa.java]: formatting

File: Essentials/src/com/earth2me/essentials/commands/Commandtpa.java
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpa.java]: Remove TODO before merge

File: Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java:43]: Not sure it should be our job to send a cancel message if another plugin is the one doing the cancelling, and likely has their own reasons for cancelling it.

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java:43]: @drtshock I think other plugins can inform players if they want, but we should definitely inform the player ourselves because this is a condition where the teleportation has failed.

**Review by mdcfe — APPROVED** [review]

**Review by drtshock — APPROVED** [review]

**Review by Ichbinjoe — APPROVED** [review]

**Pull request merged**
